Meaure auditor beta tracking performance atop pr1777

maqi opened the pull request:
### Description

Please provide a brief description of the changes made in this pull request. Highlight the purpose of the changes and the problem they address.

### Related Issue

Fixes #<issue_number> (if applicable).

### Type of Change

Please mark the types of changes made in this pull request.

<!--
### Footer (Hidden from GitHub view)
This template uses [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/). Please ensure your commit messages follow these guidelines for clarity and consistency.

Commit messages should be verified using [commitlint](https://commitlint.js.org/#/).

Commit Message Format:
<type>[optional scope]: <description>
[optional body]
[optional footer(s)]

Common Types:
- `feat`
- `fix`
- `docs`
- `style`
- `refactor`
- `perf`
- `test`
- `build`
- `ci`
- `chore`
- `revert`
-->

davidrusu commented on sn_interface/src/network_knowledge/mod.rs:
```diff
@@ -152,19 +152,28 @@ pub struct NetworkKnowledge {
+    /// Error if the SAP is not present in the tree.
     pub fn new(
-        section_tree: SectionTree,
-        section_tree_update: SectionTreeUpdate,
+        signed_sap: SectionSigned<SectionAuthorityProvider>,
+        section_tree: Option<SectionTree>,
     ) -> Result<Self, Error> {
```
This change has some non-intuitive semantics, it has two modes
1. normal case: you pass both a signed sap and a section tree. 
2. genesis case: you pass a signed_sap.

This is not ideal from an API point of view. i.e. it's not clear just from reading the following code that the signed sap is expected to be a genesis sap:

```
NetworkKnowledge::new(sap, None)
```

There's a few alternatives I think would be better:

1. Divide out the two modes into two constructors:
```rust
NetworkKnowledge::genesis(sap: SectionSigned<SAP>)
NetworkKnowledge::new(sap: SectionSigned<SAP>, section_tree: SectionTree)
```

2. take a prefix instead of sap, this makes the user constructor the section tree and pass a prefix as  a reference to where they are in the tree:
```rust
NetworkKnowledge::new(prefix: Prefix, section_tree: SectionTree)
```

I kind of like option 2. The benefits are simpler logic here in the constructor with less error checking:
1. we don't have to check that the signed sap we fetch with `tree.get_signed(prefix)` matches the signed sap passed to the constructor.
2. we don't have to check that the signed sap is a valid genesis sap (if we are in the genesis mode)
3. probably less cloning on the callers side.

RolandSherwin commented:
bors try

davidrusu commented on sn_interface/src/network_knowledge/mod.rs:
```diff
@@ -152,20 +152,13 @@ pub struct NetworkKnowledge {
+    /// Creates a new `NetworkKnowledge` instance. The prefix is used to choose the
+    /// `signed_sap` from the provided SectionTree
+    pub fn new(prefix: Prefix, section_tree: SectionTree) -> Result<Self, Error> {
+        let signed_sap = match section_tree.get_signed(&prefix) {
+            Some(signed_sap) => signed_sap.clone(),
+            None => return Err(Error::NoMatchingSection),
+        };
```
```suggestion
        let signed_sap = section_tree.get_signed(&prefix).cloned().ok_or(Error::NoMatchingSection)?;
```

davidrusu commented on sn_interface/src/network_knowledge/section_tree/mod.rs:
```diff
@@ -47,12 +47,17 @@ pub struct SectionTree {
-        Self {
-            sections: BTreeMap::new(),
-            sections_dag: SectionsDAG::new(genesis_pk),
+    /// Create a `SectionTree` with just the genesis section
+    pub fn new(genesis_sap: SectionSigned<SectionAuthorityProvider>) -> Result<Self> {
+        if genesis_sap.prefix() != Prefix::default() {
+            return Err(Error::NonGenesisSap);
         }
+
+        let genesis_pk = genesis_sap.public_key_set().public_key();
+        Ok(Self {
+            sections: BTreeMap::from([(Prefix::default(), genesis_sap)]),
+            sections_dag: SectionsDAG::new(genesis_pk),
+        })
```
👍

davidrusu reviewed (approved):
Looks good, comments are nits 🚢

RolandSherwin commented:
bors try

RolandSherwin commented:
bors try

RolandSherwin commented:
bors r+

RolandSherwin commented:
bors cancel

RolandSherwin commented:
bors r+

RolandSherwin commented:
bors r+

maqi committed:
chore(auditor): refactor spends fetch to be background progressing

maqi committed:
chore(auditor): do not track royalty and payment_forward utxo

joshuef committed:
feat(audit): use BTreeSet for participants

To avoid duplication

joshuef committed:
feat(audit): parallelise queries

joshuef committed:
HACK:fix protocol version

joshuef committed:
feat(audit): enable url decoding

fix: #1778

joshuef committed:
chore(audit): refactor and cleanup code

Address feedback, remove comments cleaner set collection and small await if nothing to scan

maqi committed:
feat(auditor): record beta tracking performance

maqi commented:
Replaced by https://github.com/maidsafe/safe_network/pull/1853
